internal/analytics/deferred.go

```diff
@@ -98,69 +104,91 @@ type Configurable interface {
 
 func deferEvent(cfg Configurable, category, action, label string, dimensions map[string]string) error {
 	logging.Debug("Deferring: %s, %s, %s", category, action, label)
-	deferred, err := loadDeferred(cfg)
-	if err != nil {
-		return errs.Wrap(err, "Could not load events on defer")
-	}
 
 	if !fileutils.FileExists(deferrerFilePath(cfg)) {
-		err = fileutils.Touch(deferrerFilePath(cfg))
-		if err != nil {
+		if err := fileutils.Touch(deferrerFilePath(cfg)); err != nil {
 			logging.Errorf("Failed to create deferrer time stamp file: %v", err)
 		}
 	}
 
-	deferred = append(deferred, deferredData{category, action, label, dimensions})
-	if err := saveDeferred(cfg, deferred); err != nil {
+	if err := saveDeferred(cfg, deferredData{category, action, label, dimensions}); err != nil {
 		return errs.Wrap(err, "Could not save event on defer")
 	}
 	return nil
 }
 
-func sendDeferred(cfg Configurable, sender func(string, string, string, map[string]string) error) error {
-	deferred, err := loadDeferred(cfg)
+func loadDeferred(deferredDataPath string) ([]deferredData, error) {
+	b, err := os.ReadFile(deferredDataPath)
 	if err != nil {
-		return errs.Wrap(err, "Could not load events on send")
+		return nil, errs.Wrap(err, "Failed to read deferred_data")
 	}
-	if len(deferred) > 0 {
-		for n, event := range deferred {
-			if err := sender(event.Category, event.Action, event.Label, event.Dimensions); err != nil {
-				return errs.Wrap(err, "Could not send deferred event")
-			}
-			if err := saveDeferred(cfg, deferred[n+1:]); err != nil {
-				return errs.Wrap(err, "Could not save deferred event on send")
+	lines := strings.Split(string(b), "\n")
+	var events []deferredData
+	var unmarshalErrorReported bool
+	for _, line := range lines {
+		var event deferredData
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+		if err := json.Unmarshal([]byte(line), &event); err != nil {
+			if !unmarshalErrorReported {
+				logging.Error("Failed to unmarshal line in deferred_data file: %v", err)
+				unmarshalErrorReported = true
 			}
+			continue
+		}
+		events = append(events, event)
+	}
+
+	return events, nil
+}
+
+func sendDeferred(cfg Configurable, sender func(string, string, string, map[string]string) error) error {
+	// move deferred data file, so it is not being appended anymore
+	outboxFile := filepath.Join(cfg.ConfigPath(), "deferred.outbox")
```

Comment:
This could still be a problem if two processes both try to flush the deferred analytics at the same time. I think we can just move it to a temp file that is uniquely named (ie. use a timestamp)